### PR TITLE
Gesture toggling

### PR DESCRIPTION
**Pull request by Zwatotem**
~~On Android so far \(except defaults for some reason\)~~

### CODE REVIEW

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 1]
```diff
@@ -9,7 +9,7 @@ import kotlinx.coroutines.*
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.collectLatest
 
-class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallback {
+class GoogleMapsView(context: Context, private val gestures: MapsViewGestures?): LinearLayout(context), OnMapReadyCallback {
 
   private val mapView: MapView = MapView(context)
   private lateinit var googleMap: GoogleMap
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]: In my opinion there is no need to pass gestures object through constructor. We can just make it class field as we did with other values. I would also use `lateinit` variable to stick to the convention we used before. Additionally I think we can type gestures used here in `GoogleMapsView` as `GoogleMapsGestures`, because this is specific implementation of gestures we have for Google Maps. What are your thoughts?
```suggestion
class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallback {

  private val mapView: MapView = MapView(context)
  private lateinit var googleMap: GoogleMap
  private lateinit var gestures: GoogleMapsGestures
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 2]
```diff
@@ -27,6 +27,10 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
 
   override fun onMapReady(googleMap: GoogleMap) {
     this.googleMap = googleMap
+    // Possibly shift this logic to a dedicated method, when there is more dependencies
+    if(gestures != null) {
+      (gestures as GoogleMapsViewGestures).initMap(googleMap)
+    }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]: As far as I understand this will be run only once (?) so we can just put constructor here and pass `googleMap` as its param. What do you think? Please verify if you understand this doc same way as I do.

CC: @p-karcz 

https://developers.google.com/android/reference/com/google/android/gms/maps/OnMapReadyCallback

```suggestion
    gestures = GoogleMapsGestures(googleMap)
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 3]
```diff
@@ -45,6 +49,36 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
     }
   }
 
+  fun setEnabledScrollGestures(enabled: Boolean) {
+    updateMap {
+      gestures?.setEnabledScrollGesture(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]:
```suggestion
      gestures.setEnabledScrollGesture(enabled)
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 4]
```diff
@@ -45,6 +49,36 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
...
+    }
+  }
+
+  fun setEnabledZoomGestures(enabled: Boolean) {
+    updateMap {
+      gestures?.setEnabledZoomGesture(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]:
```suggestion
      gestures.setEnabledZoomGesture(enabled)
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 5]
```diff
@@ -45,6 +49,36 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
...
+    }
+  }
+
+  fun setEnabledTiltGestures(enabled: Boolean) {
+    updateMap {
+      gestures?.setEnabledTiltGesture(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]:
```suggestion
      gestures.setEnabledTiltGesture(enabled)
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 6]
```diff
@@ -45,6 +49,36 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
...
+    }
+  }
+
+  fun setEnabledRotateGestures(enabled: Boolean) {
+    updateMap {
+      gestures?.setEnabledRotateGesture(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]:
```suggestion
      gestures.setEnabledRotateGesture(enabled)
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 7]
```diff
@@ -45,6 +49,36 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
...
+    }
+  }
+
+  fun setEnabledAllGestures(enabled: Boolean) {
+    updateMap {
+      gestures?.setEnabledAllGestures(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]:
```suggestion
      gestures.setEnabledAllGestures(enabled)
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.google.android.gms.maps.GoogleMap
+
+class GoogleMapsViewGestures() : MapsViewGestures {
+
+  private var googleMap: GoogleMap? = null
+
+  fun initMap(googleMap: GoogleMap) {
+    this.googleMap = googleMap
+  }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt]: I think we named markers class `GoogleMapsMarkers` so we can try to keep it same way, although this is very subjective and isn’t issue at all. What do you prefer to do?

We can also try to initialize `googleMap` object using constructor as I proposed above.

```suggestion
class GoogleMapsGestures(map: GoogleMap) : Gestures {

  private var googleMap: GoogleMap = map
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+  override fun setEnabledRotateGesture(enabled: Boolean) {
+    googleMap?.uiSettings?.isRotateGesturesEnabled = enabled
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt]:
```suggestion
    googleMap.uiSettings.isRotateGesturesEnabled = enabled
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  override fun setEnabledScrollGesture(enabled: Boolean) {
+    googleMap?.uiSettings?.isScrollGesturesEnabled = enabled
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt]:
```suggestion
    googleMap.uiSettings.isScrollGesturesEnabled = enabled
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  override fun setEnabledTiltGesture(enabled: Boolean) {
+    googleMap?.uiSettings?.isTiltGesturesEnabled = enabled
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt]:
```suggestion
    googleMap.uiSettings.isTiltGesturesEnabled = enabled
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  override fun setEnabledZoomGesture(enabled: Boolean) {
+    googleMap?.uiSettings?.isZoomGesturesEnabled = enabled
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt]:
```suggestion
    googleMap.uiSettings.isZoomGesturesEnabled = enabled
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  override fun setEnabledAllGestures(enabled: Boolean) {
+    googleMap?.uiSettings?.setAllGesturesEnabled(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt]:
```suggestion
    googleMap.uiSettings.setAllGesturesEnabled(enabled)
```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  }
+}
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsViewGestures.kt]:
```suggestion
}

```

File: expo-maps/android/src/main/java/expo/modules/maps/MapsViewGestures.kt [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package expo.modules.maps
+
+interface MapsViewGestures {
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/MapsViewGestures.kt]: We named markers interface `Markers` so to keep it consistent we can name this `Gestures`, although it is very subjective thing and we can decide other way. Let me know which name is better in your opinion.

```suggestion
interface Gestures {
```

File: expo-maps/android/src/main/java/expo/modules/maps/MapsViewGestures.kt [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+  fun setEnabledRotateGesture(enabled: Boolean)
+  fun setEnabledScrollGesture(enabled: Boolean)
+  fun setEnabledTiltGesture(enabled: Boolean)
+  fun setEnabledZoomGesture(enabled: Boolean)
+  fun setEnabledAllGestures(enabled: Boolean)
+}
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/MapsViewGestures.kt]:
```suggestion
}

```

File: expo-maps/src/Maps.types.ts [hunk 17]
```diff
@@ -15,4 +19,8 @@ export type ExpoMapViewProps = ViewProps &
   PropsWithChildren<{
     provider?: 'google' | 'apple';
     mapType?: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    enableRotateGestures: boolean;
+    enableScrollGestures: boolean;
+    enableTiltGestures: boolean;
+    enableZoomGestures: boolean;
```
Comment (szdziedzic) [inline expo-maps/src/Maps.types.ts]: I think we can make these props optional since we use `defaultNativeExpoMapViewProps`, so user isn’t forced to specify them while using our library.

```suggestion
    enableRotateGestures?: boolean;
    enableScrollGestures?: boolean;
    enableTiltGestures?: boolean;
    enableZoomGestures?: boolean;
```

**Review by szdziedzic — COMMENTED** [review]
Good job 👍. Overall everything seems good. I have few questions / comments / suggestions for you:

Also one small tip: don’t forget to run `yarn build` command in `expo-maps/expo-maps` folder once you change `Expo Maps` typescript files 😀

File: expo-maps/android/src/main/java/expo/modules/maps/MapsViewGestures.kt (hunk 15 above)
Comment (Zwatotem) [reply expo-maps/android/src/main/java/expo/modules/maps/MapsViewGestures.kt]: I don't mind renaming this. I just wasn't aware of @ellwiii's PR at the time of writing this.

File: expo-maps/src/Maps.types.ts (hunk 17 above)
Comment (Zwatotem) [reply expo-maps/src/Maps.types.ts]: Sure, arriving in next commit.

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt (hunk 1 above)
Comment (Zwatotem) [reply expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]: All the other things you commented on are done the way they are because I wanted this injection through constructor. But you may be right, there is not much of implementation freedom to classes like GoogleMapsGestures, that would make anyone want to replace it with something different. In return we gain more clarity about the way things are instantiated.

**Comment by Zwatotem** [comment]
> Good job 👍. Overall everything seems good. I have few questions / comments / suggestions for you:
> 
> Also one small tip: don’t forget to run `yarn build` command in `expo-maps/expo-maps` folder once you change `Expo Maps` typescript files 😀 

Yes, I realized that soon after. This is a draft for a reason 😅

File: example/ios/Podfile.lock [hunk 18]
```diff
@@ -512,4 +512,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: c8aa0d312be23a613d8cec127cd5d99f85141dfa
 
-COCOAPODS: 1.10.1
```
Comment (lukmccall) [inline example/ios/Podfile.lock]: As a team, you should get along about one specific cocoapods version. Different versions can generate sometimes different `Podfile.lock` and that will generate more unnecessary changes

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsGestures.kt [hunk 19]
```diff
@@ -0,0 +1,26 @@
+package expo.modules.maps
+
+import com.google.android.gms.maps.GoogleMap
+
+class GoogleMapsGestures(private val googleMap: GoogleMap) : Gestures {
+
```
Comment (lukmccall) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsGestures.kt:6]:
```suggestion
val uiSettings get() = googleMap.uiSettings
```
I would create a getter for UI settings to not write `googleMap.uiSettings` every time. But it's up to you.

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 20]
```diff
@@ -27,6 +28,7 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
 
   override fun onMapReady(googleMap: GoogleMap) {
     this.googleMap = googleMap
+    this.gestures = GoogleMapsGestures(this.googleMap)
```
Comment (lukmccall) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]:
```suggestion
   gestures = GoogleMapsGestures(this.googleMap)
```

File: expo-maps/ios/ExpoMaps/AppleMapsGestures.swift [hunk 21]
```diff
@@ -0,0 +1,40 @@
+import MapKit
+final class AppleMapsGestures : Gestures
+{
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsGestures.swift]:
```suggestion
final class AppleMapsGestures : Gestures {
```

File: expo-maps/ios/ExpoMaps/AppleMapsGestures.swift [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+  
+  internal var mapView: MKMapView;
+  
+  init(mapView: MKMapView)
+  {
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsGestures.swift]:
```suggestion
  init(mapView: MKMapView) {
```

File: expo-maps/ios/ExpoMaps/AppleMapsGestures.swift [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+  }
+  
+  func setEnabledAllGestures(enabled: Bool) {
+    mapView.isRotateEnabled = enabled
+    mapView.isPitchEnabled = enabled
+    mapView.isScrollEnabled = enabled
+    mapView.isZoomEnabled = enabled
+  }
+    
+
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsGestures.swift]:
```suggestion
```

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 24]
```diff
@@ -2,10 +2,12 @@ import MapKit
 
 public final class AppleMapsView: UIView {
   internal var mapView: MKMapView
+  internal var gestures: AppleMapsGestures
 
   init() {
     self.mapView = MKMapView()
     self.mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+    self.gestures = AppleMapsGestures(mapView: self.mapView)
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
   gestures = AppleMapsGestures(mapView: self.mapView)
```

File: expo-maps/ios/ExpoMaps/Gestures.swift [hunk 25]
```diff
@@ -0,0 +1,10 @@
+protocol Gestures
+{
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/Gestures.swift]:
```suggestion
protocol Gestures {    
```

File: expo-maps/ios/ExpoMaps/GoogleMapsGestures.swift [hunk 26]
```diff
@@ -0,0 +1,35 @@
+import GoogleMaps
+
+final class GoogleMapsGestures : Gestures
+{
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/GoogleMapsGestures.swift]:
```suggestion
final class GoogleMapsGestures : Gestures {
```

File: expo-maps/ios/ExpoMaps/GoogleMapsGestures.swift [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+    internal var mapView: GMSMapView
+    
+    init(mapView: GMSMapView)
+    {
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/GoogleMapsGestures.swift]:
```suggestion
    init(mapView: GMSMapView) {
```

**Review by lukmccall — APPROVED** [review]
LGTM ✅

**Review by szdziedzic — APPROVED** [review]
Good job 👏! LGTM as well 😀
